### PR TITLE
clp-s: Add archive_id to search results.

### PR DESCRIPTION
**Pull request by haiqi96**
# References

# Description
This PR adds archive_id to the search results of clp-s and a few refactorings:
1. Let clp-s:ArchiveReader tracks archive_id and update the signature of open() function.
2. replace std::string with std::string_view in clp-s' OutputHandler
3. replace "should_output_timestamp" with "should_output_metadata"

# Validation performed
1. Compressed and searched mongod log with clp-s binary directly. Verified that results are output properly
2. Ingested mongod log with clp-s package. Verified that both search.sh and webui can return the results properly
3. Confirmed that messages in the results_cache have associated archive_id.

### CODE REVIEW

File: components/core/src/clp_s/ArchiveReader.cpp [hunk 1]
```diff
@@ -1,15 +1,23 @@
...
+
 namespace clp_s {
-void ArchiveReader::open(std::string const& archive_path) {
+void ArchiveReader::open(string_view archives_dir, string_view archive_id) {
     if (m_is_open) {
         throw OperationFailed(ErrorCodeNotReady, __FILENAME__, __LINE__);
     }
     m_is_open = true;
-    m_archive_path = archive_path;
+    m_archive_id = archive_id;
```
Comment (kirkrodrigues) [inline components/core/src/clp_s/ArchiveReader.cpp:17]: If I'm not mistaken, `m_archive_id` is only used in this method, so we could make it a local variable.

File: components/core/src/clp_s/clp-s.cpp [hunk 2]
```diff
@@ -346,8 +344,8 @@ int main(int argc, char const* argv[]) {
                     // Skip non-directories
                     continue;
                 }
-
-                archive_reader->open(entry.path());
+                auto const archive_id = entry.path().filename().string();
```
Comment (kirkrodrigues) [inline components/core/src/clp_s/clp-s.cpp:347]:
```suggestion

                auto const archive_id = entry.path().filename().string();
```

File: components/core/src/clp_s/clp-s.cpp [hunk 3]
```diff
@@ -271,10 +271,11 @@ int main(int argc, char const* argv[]) {
 
         clp_s::JsonConstructorOption option;
         option.output_dir = command_line_arguments.get_output_dir();
+        option.archives_dir = std::filesystem::path{archives_dir};
```
Comment (kirkrodrigues) [inline components/core/src/clp_s/clp-s.cpp:274]: Why turn this into a filesystem path? `option.archives_dir` is a string. Did you want to turn `option.archives_dir` and the parameter to `ArchiveReader::open` into a filesystem path?

File: components/core/src/clp_s/search/OutputHandler.cpp [hunk 4]
```diff
@@ -23,18 +28,14 @@ NetworkOutputHandler::NetworkOutputHandler(
...
-void NetworkOutputHandler::write(std::string const& message) {
-    msgpack::type::tuple<epochtime_t, std::string, std::string> src(0, message, "");
+void NetworkOutputHandler::write(
+        string_view message,
+        epochtime_t timestamp,
+        string_view archive_id
+) {
+    static constexpr string_view cOrigFilePathPlaceholder{""};
+    msgpack::type::tuple<epochtime_t, string, string, string> const
+            src(timestamp, message, cOrigFilePathPlaceholder, archive_id);
```
Comment (kirkrodrigues) [inline components/core/src/clp_s/search/OutputHandler.cpp:38]:
```suggestion
    msgpack::type::tuple<epochtime_t, string, string, string> const
            src(timestamp, message, string_view{}, archive_id);
```

File: components/core/src/clp_s/search/OutputHandler.cpp [hunk 5]
```diff
@@ -98,12 +100,16 @@ ErrorCode ResultsCacheOutputHandler::flush() {
...
 
-void ResultsCacheOutputHandler::write(std::string const& message, epochtime_t timestamp) {
+void ResultsCacheOutputHandler::write(
+        string_view message,
+        epochtime_t timestamp,
+        string_view archive_id
+) {
     if (m_latest_results.size() < m_max_num_results) {
-        m_latest_results.emplace(std::make_unique<QueryResult>("", message, timestamp));
+        m_latest_results.emplace(std::make_unique<QueryResult>("", message, timestamp, archive_id));
```
Comment (kirkrodrigues) [inline components/core/src/clp_s/search/OutputHandler.cpp:109]:
```suggestion
        m_latest_results.emplace(std::make_unique<QueryResult>(string_view{}, message, timestamp, archive_id));
```

File: components/core/src/clp_s/search/OutputHandler.cpp [hunk 6]
```diff
@@ -98,12 +100,16 @@ ErrorCode ResultsCacheOutputHandler::flush() {
...
     } else if (m_latest_results.top()->timestamp < timestamp) {
         m_latest_results.pop();
-        m_latest_results.emplace(std::make_unique<QueryResult>("", message, timestamp));
+        m_latest_results.emplace(std::make_unique<QueryResult>("", message, timestamp, archive_id));
```
Comment (kirkrodrigues) [inline components/core/src/clp_s/search/OutputHandler.cpp:112]:
```suggestion
        m_latest_results.emplace(std::make_unique<QueryResult>(string_view{}, message, timestamp, archive_id));
```

Comment (kirkrodrigues) [reply components/core/src/clp_s/search/OutputHandler.cpp:112]: Apply linter after.

File: components/core/src/clp_s/search/OutputHandler.hpp [hunk 7]
```diff
@@ -115,9 +119,10 @@ class NetworkOutputHandler : public OutputHandler {
     }
 
     // Methods inherited from OutputHandler
-    void write(std::string const& message, epochtime_t timestamp) override;
+    void
+    write(std::string_view message, epochtime_t timestamp, std::string_view archive_id) override;
 
-    void write(std::string const& message) override;
+    void write(std::string_view message) override { write(message, 0, ""); }
```
Comment (kirkrodrigues) [inline components/core/src/clp_s/search/OutputHandler.hpp:125]:
```suggestion
    void write(std::string_view message) override { write(message, 0, {}); }
```

File: components/core/src/clp_s/search/OutputHandler.hpp [hunk 8]
```diff
@@ -176,9 +188,10 @@ class ResultsCacheOutputHandler : public OutputHandler {
      */
     ErrorCode flush() override;
 
-    void write(std::string const& message, epochtime_t timestamp) override;
+    void
+    write(std::string_view message, epochtime_t timestamp, std::string_view archive_id) override;
 
-    void write(std::string const& message) override { write(message, 0); }
+    void write(std::string_view message) override { write(message, 0, ""); }
```
Comment (kirkrodrigues) [inline components/core/src/clp_s/search/OutputHandler.hpp:194]:
```suggestion
    void write(std::string_view message) override { write(message, 0, {}); }
```

**Review by kirkrodrigues — CHANGES_REQUESTED** [review]

File: components/core/src/clp_s/search/OutputHandler.cpp (hunk 4 above)
Comment (haiqi96) [reply components/core/src/clp_s/search/OutputHandler.cpp:38]: I personally would prefer to give the 3rd parameter a name, so it's clear for developers to see what has been assigned an empty string.

File: components/core/src/clp_s/ArchiveReader.cpp (hunk 1 above)
Comment (haiqi96) [reply components/core/src/clp_s/ArchiveReader.cpp:17]: we do have this method though

https://github.com/y-scope/clp/pull/435/files/674f8b414292b0bc6e94f1a7ba0e485efa134663#diff-aab9a16ec680184cc79784526e0e20dd1b8d0ad18737eda5f9a31dc1ee7bc415R97

Comment (kirkrodrigues) [reply components/core/src/clp_s/ArchiveReader.cpp:17]: Sorry, I mean `m_archive_path`.

File: components/core/src/clp_s/search/OutputHandler.cpp (hunk 4 above)
Comment (kirkrodrigues) [reply components/core/src/clp_s/search/OutputHandler.cpp:38]: Sure.

File: components/core/src/clp_s/search/Output.cpp [hunk 9]
```diff
@@ -85,15 +85,15 @@ bool Output::filter() {
...
                 m_should_marshal_records
         );
         reader.initialize_filter(this);
 
-        if (m_output_handler->should_output_timestamp()) {
+        if (m_output_handler->should_output_metadata()) {
             epochtime_t timestamp;
             while (reader.get_next_message_with_timestamp(message, timestamp, this)) {
-                m_output_handler->write(message, timestamp);
+                m_output_handler->write(message, timestamp, m_archive_reader->get_archive_id());
```
Comment (wraymo) [inline components/core/src/clp_s/search/Output.cpp:96]: Do we want to extract `m_archive_reader->get_archive_id()` out since it doesn't change?

Comment (haiqi96) [reply components/core/src/clp_s/search/Output.cpp]: Thanks for catching it. fixed

File: components/core/src/clp_s/search/Output.cpp [hunk 10]
```diff
@@ -65,6 +65,7 @@ bool Output::filter() {
     populate_string_queries(top_level_expr);
 
     std::string message;
+    auto const& archive_id = m_archive_reader->get_archive_id();
```
Comment (kirkrodrigues) [inline components/core/src/clp_s/search/Output.cpp:68]:
```suggestion
    auto const archive_id = m_archive_reader->get_archive_id();
```

It's a bit weird to use a reference for an rvalue.

File: components/core/src/clp_s/ArchiveReader.cpp [hunk 11]
```diff
@@ -1,26 +1,34 @@
...
+    std::filesystem::path archive_path{archives_dir};
+    archive_path /= m_archive_id;
+    auto const& archive_path_str = archive_path.string();
```
Comment (kirkrodrigues) [inline components/core/src/clp_s/ArchiveReader.cpp:20]:
```suggestion
    auto const archive_path_str = archive_path.string();
```

**Review by kirkrodrigues — CHANGES_REQUESTED** [review]

**Review by kirkrodrigues — APPROVED** [review]
For the PR title, how about:

clp-s: Add archive_id to search results.

**Pull request merged**